File: Conversor/hexconverter.go

package main

import (
    "encoding/hex"
    "fmt"
    "os"
    "strings"
)

func main() {
    // Exemplo de pacote IP com payload | CONTÉM PAYLOAD "HELLO WORLD"
    hexString := "45 00 00 3C 1C 46 40 00 40 06 B1 E6 C0 A8 01 02 C0 A8 01 01 48 65 6C 6C 6F 20 57 6F 72 6C 64"


    hexString = strings.ReplaceAll(hexString, " ", "")
    hexString = strings.ReplaceAll(hexString, "-", "")

    data, err := hex.DecodeString(hexString)
    if err != nil {
        fmt.Println("Erro ao converter hex:", err)
        os.Exit(1)
    }

    if len(data) < 20 {
        fmt.Println("Pacote inválido: cabeçalho IPv4 precisa de pelo menos 20 bytes.")
        os.Exit(1)
    }

    version := data[0] >> 4
    headerLength := (data[0] & 0x0F) * 4
    totalLength := uint16(data[2])<<8 | uint16(data[3])
    identification := uint16(data[4])<<8 | uint16(data[5])
    flags := data[6] >> 5
    fragmentOffset := (uint16(data[6]&0x1F) << 8) | uint16(data[7])
    ttl := data[8]
    protocol := data[9]
    checksum := uint16(data[10])<<8 | uint16(data[11])
    sourceIP := fmt.Sprintf("%d.%d.%d.%d", data[12], data[13], data[14], data[15])
    destIP := fmt.Sprintf("%d.%d.%d.%d", data[16], data[17], data[18], data[19])

    fmt.Println("========= Cabeçalho IP =========")
    fmt.Printf("Version: %d\n", version)
    fmt.Printf("Header Length: %d bytes\n", headerLength)
    fmt.Printf("Total Length: %d\n", totalLength)
    fmt.Printf("Identification: %d\n", identification)
    fmt.Printf("Flags: %d\n", flags)
    fmt.Printf("Fragment Offset: %d\n", fragmentOffset)
    fmt.Printf("TTL: %d\n", ttl)
    fmt.Printf("Protocol: %d\n", protocol)
    fmt.Printf("Header Checksum: %d\n", checksum)
    fmt.Printf("Source IP: %s\n", sourceIP)
    fmt.Printf("Destination IP: %s\n", destIP)
    fmt.Println("================================")

    // Payload
    if len(data) > int(headerLength) {
        payload := data[headerLength:]
        ascii := ""
        for _, b := range payload {
            if b >= 32 && b <= 126 {
                ascii += string(b)
            } else {
                ascii += "."
            }
        }
        fmt.Printf("Payload ASCII: %s\n", ascii)
    } else {
        fmt.Println("Sem payload no pacote.")
    }
}

File: Conversor/hexconverter.py
def dissect_ip_packet(hex_string):
    # Remove spaces and hyphens
    hex_string = hex_string.replace(" ", "").replace("-", "")
    
    # Convert hex to decimal
    decimal_values = [int(hex_string[i:i+2], 16) for i in range(0, len(hex_string), 2)]
    
    # Dissect the IP packet
    ip_packet = {
        "Version": (decimal_values[0] >> 4),
        "Header Length": (decimal_values[0] & 0x0F) * 4,
        "Explicit Congestion Notification": decimal_values[1],
        "Total Length": (decimal_values[2] << 8) + decimal_values[3],
        "Identification": (decimal_values[4] << 8) + decimal_values[5],
        "Flags": (decimal_values[6] >> 5),
        "Fragment Offset": ((decimal_values[6] & 0x1F) << 8) + decimal_values[7],
        "TTL": decimal_values[8],
        "Protocol": decimal_values[9],
        "Header Checksum": (decimal_values[10] << 8) + decimal_values[11],
        "Source IP": f"{decimal_values[12]}.{decimal_values[13]}.{decimal_values[14]}.{decimal_values[15]}",
        "Destination IP": f"{decimal_values[16]}.{decimal_values[17]}.{decimal_values[18]}.{decimal_values[19]}"
    }
    
    return ip_packet

def extract_ascii(hex_string):
    # Remove spaces and hyphens
    hex_string = hex_string.replace(" ", "").replace("-", "")
    
    # Convert hex to decimal
    decimal_values = [int(hex_string[i:i+2], 16) for i in range(0, len(hex_string), 2)]
    
    # Convert decimal to ASCII characters
    ascii_characters = [chr(value) for value in decimal_values]
    
    # Join ASCII characters to form the plain text
    plain_text = ''.join(ascii_characters)
    
    return plain_text

# Hexadecimal string example
hex_string = "45 00 00 3C 1C 46 40 00 40 06 B1 E6 C0 A8 01 02 C0 A8 01 01F"

# Dissect the IP packet and print the results
ip_packet = dissect_ip_packet(hex_string)
print("========= Pacote IP ============")
for field, value in ip_packet.items():
    print(f"{field}: {value}")
print("===============================")

# Extract ASCII and print the results
plain_text = extract_ascii(hex_string)
print(f"ASCII Text: {plain_text}")
